### ISSUE TITLE
Assert fails in Conv.CheckInvariant: incompatible conv instruction remains after ldobj(ldloca)->ldloc transform involving unknown types

### ISSUE THREAD

**Issue by greenozon** [icsharpcode/ILSpy#1366]
ILSpy version 4.0.0.4482-beta3

Note: might be a dup, need analysis

```
Error decompiling <Module>.PCTimeDimGenerator.CreateTimeDimDef
 ---> ICSharpCode.ILSpy.AssertionFailedException: 
   at ICSharpCode.ILSpy.ILSpyTraceListener.Fail(String message, String detailMessage) in offset 380
   at ICSharpCode.ILSpy.ILSpyTraceListener.Fail(String message) in offset 9
   at System.Diagnostics.TraceInternal.Fail(String message) in offset 55
   at System.Diagnostics.Debug.Assert(Boolean condition) in offset 6
   at ICSharpCode.Decompiler.IL.Conv.CheckInvariant(ILPhase phase) in offset 44
   at ICSharpCode.Decompiler.IL.ILInstruction.CheckInvariant(ILPhase phase) in offset 126
   at ICSharpCode.Decompiler.IL.CallInstruction.CheckInvariant(ILPhase phase) in offset 8
   at ICSharpCode.Decompiler.IL.ILInstruction.CheckInvariant(ILPhase phase) in offset 126
   at ICSharpCode.Decompiler.IL.Block.CheckInvariant(ILPhase phase) in offset 8
   at ICSharpCode.Decompiler.IL.ILInstruction.CheckInvariant(ILPhase phase) in offset 126
   at ICSharpCode.Decompiler.IL.BlockContainer.CheckInvariant(ILPhase phase) in offset 8
   at ICSharpCode.Decompiler.IL.ILInstruction.CheckInvariant(ILPhase phase) in offset 126
   at ICSharpCode.Decompiler.IL.ILFunction.CheckInvariant(ILPhase phase) in offset 88
   at ICSharpCode.Decompiler.CSharp.CSharpDecompiler.DecompileBody(IMethod method, EntityDeclaration entityDecl, DecompileRun decompileRun, ITypeResolveContext decompilationContext) in offset 590
-- continuing with outer exception (ICSharpCode.Decompiler.DecompilerException) --
   at ICSharpCode.Decompiler.CSharp.CSharpDecompiler.DecompileBody(IMethod method, EntityDeclaration entityDecl, DecompileRun decompileRun, ITypeResolveContext decompilationContext) in offset 1068
   at ICSharpCode.Decompiler.CSharp.CSharpDecompiler.DoDecompile(IMethod method, DecompileRun decompileRun, ITypeResolveContext decompilationContext) in offset 249
   at ICSharpCode.Decompiler.CSharp.CSharpDecompiler.Decompile(IList`1 definitions) in offset 408
   at ICSharpCode.Decompiler.CSharp.CSharpDecompiler.Decompile(EntityHandle[] definitions) in offset 1
   at ICSharpCode.ILSpy.CSharpLanguage.DecompileMethod(IMethod method, ITextOutput output, DecompilationOptions options) in offset 246
   at ICSharpCode.ILSpy.TreeNodes.MethodTreeNode.Decompile(Language language, ITextOutput output, DecompilationOptions options) in offset 15
   at ICSharpCode.ILSpy.TextView.DecompilerTextView.DecompileNodes(DecompilationContext context, ITextOutput textOutput) in offset 69
   at ICSharpCode.ILSpy.TextView.DecompilerTextView.<>c__DisplayClass34_0.<DecompileAsync>b__0() in offset 38

```

target: https://www.sendspace.com/file/56em1a

**Comment by siegfriedpammer** [comment]
Sorry, I forgot to download the binary. Could you please upload it again? Thank you very much!

**Comment by greenozon** [comment]
[err40.zip](https://github.com/icsharpcode/ILSpy/files/3389908/err40.zip)


PS if you save proj, it'll generate another assert, but if you go to above method you'll have reproduce

**Comment by siegfriedpammer** [comment]
ILAst before the assertion:
```
stloc V_16(conv i4->unknown <invalid>(ldloc S_89))
call set_FiscalYearName(conv o->unknown <invalid>(ldloc V_0), conv i4->unknown <invalid>(ldobj System.Int32(ldloca V_16)))
```

ILAst after `ldobj(ldloca V_16) => ldloc V_16`:
```
stloc V_16(conv i4->unknown <invalid>(ldloc S_89))
call set_FiscalYearName(conv o->unknown <invalid>(ldloc V_0), conv i4->unknown <invalid>(ldloc V_16))
```

Because we replaced the `ldobj T(ldloca)` with a simple `ldloc`, the types as expected by the outer conv instruction no longer match up.

**Comment by dgrunwald** [comment]
But the `ldobj(ldloca) -> ldloc` transform is checking that the type of the local variable is compatible with the type used for the load (`IsCompatibleTypeForMemoryAccess`)?

**Comment by siegfriedpammer** [comment]
https://github.com/icsharpcode/ILSpy/blob/c148ba6609c13f3d9253f20da9ab6be13321bee6/ICSharpCode.Decompiler/TypeSystem/TypeUtils.cs#L228-L229

**Comment by siegfriedpammer** [comment]
@dgrunwald so the correct solution in this case would be to remove the outer `conv` instruction (as a conversion from `i4` to `unknown` is no longer required)?

**Comment by siegfriedpammer** [comment]
@dgrunwald any idea?

**Comment by siegfriedpammer** [comment]
Probably fixed by 15d74751f48cdb828bd4fd97a4d6af33b38eb05d
